presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyAggregation.java

```diff
@@ -0,0 +1,201 @@
+{
+    private DifferentialEntropyAggregation() {}
+
+    @InputFunction
+    public static void input(
+            @AggregationState DifferentialEntropyState state,
+            @SqlType(StandardTypes.BIGINT) long size,
+            @SqlType(StandardTypes.DOUBLE) double sample,
+            @SqlType(StandardTypes.DOUBLE) double weight,
+            @SqlType(StandardTypes.VARCHAR) Slice method,
+            @SqlType(StandardTypes.DOUBLE) double min,
+            @SqlType(StandardTypes.DOUBLE) double max)
+    {
+        String requestedMethod = method == null ? null : method.toStringUtf8();
+        if (state.getStrategy() == null) {
+            DifferentialEntropyStateStrategy strategy;
+            if (requestedMethod.equalsIgnoreCase(FIXED_HISTOGRAM_MLE_METHOD_NAME)) {
+                strategy = new FixedHistogramMLEStateStrategy(size, min, max);
+            }
+            else if (requestedMethod.equalsIgnoreCase(FIXED_HISTOGRAM_JACKNIFE_METHOD_NAME)) {
+                strategy = new FixedHistogramJacknifeStateStrategy(size, min, max);
+            }
+            else {
+                throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
+            }
+            state.setStrategy(strategy);
+        }
+        DifferentialEntropyStateStrategy strategy = state.getStrategy();
+        validate(strategy, requestedMethod);
+        strategy.validateParameters(size, sample, weight, min, max);
+        strategy.add(sample, weight);
+    }
+
+    @InputFunction
+    public static void input(
+            @AggregationState DifferentialEntropyState state,
+            @SqlType(StandardTypes.BIGINT) long size,
+            @SqlType(StandardTypes.DOUBLE) double sample,
+            @SqlType(StandardTypes.DOUBLE) double weight,
+            @SqlType(StandardTypes.VARCHAR) Slice method)
+    {
+        String requestedMethod = method.toStringUtf8();
+        if (state.getStrategy() == null) {
+            DifferentialEntropyStateStrategy strategy = new WeightedReservoirSampleStateStrategy(size);
+            state.setStrategy(strategy);
+        }
+        DifferentialEntropyStateStrategy strategy = state.getStrategy();
+        validate(strategy, requestedMethod);
+        strategy.validateParameters(size, sample, weight);
+        strategy.add(sample, weight);
+    }
+
+    @InputFunction
+    public static void input(
+            @AggregationState DifferentialEntropyState state,
+            @SqlType(StandardTypes.BIGINT) long size,
+            @SqlType(StandardTypes.DOUBLE) double sample,
+            @SqlType(StandardTypes.DOUBLE) double weight)
+    {
+        if (state.getStrategy() == null) {
+            DifferentialEntropyStateStrategy strategy = new WeightedReservoirSampleStateStrategy(size);
+            state.setStrategy(strategy);
+        }
+        DifferentialEntropyStateStrategy strategy = state.getStrategy();
+        validate(strategy);
```

Comment:
I'd remove this method and add explicit verify calls:

```
        DifferentialEntropyStateStrategy strategy = state.getStrategy();
        if (strategy == null) {
            strategy = new WeightedReservoirSampleStateStrategy(size);
            state.setStrategy(strategy);
        }
        else {
            verify(strategy instanceof WeightedReservoirSampleStateStrategy);
        }
```

Reply:
Not sure I follow. It is called from two places.

Reply:
I suggest to change both places to add explicit verify call.